morango/models.py

```diff
@@ -288,17 +288,33 @@ def _deserialize_store_model(self):
         klass_model = _profile_models[self.profile][self.model_name]
         # if store model marked as deleted, attempt to delete in app layer
         if self.deleted:
-            klass_model.objects.filter(id=self.id).delete()
+            # if hard deleted, propogate to related models
+            if self.hard_delete:
+                try:
+                    klass_model.objects.get(id=self.id).delete(hard_delete=True)
+                except klass_model.DoesNotExist:
+                    pass
+            else:
+                klass_model.objects.filter(id=self.id).delete()
         # inflate model and attempt to save
         else:
             app_model = klass_model.deserialize(json.loads(self.serialized))
             app_model._morango_source_id = self.source_id
             app_model._morango_partition = self.partition
             try:
+                app_model.clean_fields()
                 with mute_signals(signals.pre_save, signals.post_save):
                     app_model.save(update_dirty_bit_to=False)
             # if unable to save due to missing FKs, mark model as deleted
-            except ObjectDoesNotExist:
+            except ValidationError:
+                # check FKs in store to see if any of those models were hard deleted
+                fk_ids = [getattr(app_model, field.attname) for field in app_model._meta.fields if isinstance(field, ForeignKey)]
+                for fk_id in fk_ids:
+                    try:
+                        if Store.objects.get(id=fk_id).hard_delete:
+                            app_model._update_hard_deleted_models()
+                    except Store.DoesNotExist:
+                        pass
                 app_model._update_deleted_models()
```

Comment:
As discussed in a comment on the commit -- probably best to only do the auto-deletion when we explicitly see that the FK points to a deleted record in the store (can do jointly with what you do for hard-deletion, just above this). Otherwise, leaving the dirty bit set so it can try again later.